book/scheduling-and-threading.md

```diff
@@ -327,116 +275,129 @@ situation---some devices have more [CPU cores][cores] than others, or are more
 
-thread = threading.Thread(target=run)
-thread.start()
-while True:
-    print("Browser")
+``` {.javascript file=runtime}
+XHR_REQUESTS = {}
+
+function XMLHttpRequest() {
+    this.handle = Object.keys(XHR_REQUESTS).length;
+    XHR_REQUESTS[this.handle] = this;
+}
 ```
 
-a stream of lines with the words "Thread" and "Browser" will be
-interspersed---according to the CPU scheduling algorithm of the computer and
-the Python runtime---until the thread has printed 100 times, after which
-"Browser" will continue printing forever.
+When a script calls the `open` method on an `XMLHttpRequest` object,
+we'll now allow the `is_async` flag to be true:
 
-Here is the code for `XMLHttpRequest_send` (the new `is_async` parameter
-indicates an async request). In this case, `run_load` is the thread main
-function, and after the line that says "return out" executes, the thread will
-die.^[Note that for async requests, the return statement is meaningless;
-it's only there for the sync version.]
+``` {.javascript file=runtime}
+XMLHttpRequest.prototype.open = function(method, url, is_async) {
+    this.is_async = is_async
+    this.method = method;
+    this.url = url;
+}
+```
 
-``` {.python}
-XHR_ONLOAD_CODE = "__runXHROnload(dukpy.out, dukpy.handle)"
+The `send` method will need to send over the `is_async` flag and the
+handle:
 
-class JSContext:
-    def xhr_onload(self, out, handle):
-        do_default = self.interp.evaljs(
-            XHR_ONLOAD_CODE, out=out, handle=handle)
+``` {.javascript file=runtime}
+XMLHttpRequest.prototype.send = function(body) {
+    this.responseText = call_python("XMLHttpRequest_send",
+        this.method, this.url, this.body, this.is_async, this.handle);
+}
+```
+
+On the browser side, we'll need to split the `XMLHttpRequest_send`
+function into three parts. The first part will resolve the URL and
+do security checks:
 
-    def XMLHttpRequest_send(
-        self, method, url, body, is_async, handle):
+``` {.python}
+class JSContext:
+    def XMLHttpRequest_send(self, method, url, body, is_async, handle):
         full_url = resolve_url(url, self.tab.url)
         if not self.tab.allowed_request(full_url):
             raise Exception("Cross-origin XHR blocked by CSP")
         if url_origin(full_url) != url_origin(self.tab.url):
             raise Exception(
                 "Cross-origin XHR request not allowed")
+```
 
+Then, we'll define a function that makes the request and enqueues a
+task for running callbacks:^[For async requests, the `return` at the
+end of `run_load` is meaningless; it's only there for the sync version.]
+
+``` {.python}
+class JSContext:
+    def XMLHttpRequest_send(self, method, url, body, is_async, handle):
+        # ...
         def run_load():
-            headers, out = request(
+            headers, body = request(
                 full_url, self.tab.url, payload=body)
-            handle_local = handle
             self.tab.task_runner.schedule_task(
```

Comment:
I think there is my bug in the code here - should only schedule a task if is_async is true?

Reply:
You're right. Fixed. Do you think we should add some code to default `is_async` to `true` on the JavaScript side? Otherwise if you don't pass the value we will assume it is sync. Argument against is just that the code is super ugly since DukPy doesn't support ES6. Maybe we footnote it?

Reply:
Let's go with a footnote.